app/controllers/users_controller.rb

```diff
@@ -496,15 +496,29 @@ def is_local_username
     usernames -= groups
     usernames.each(&:downcase!)
 
-    cannot_see = []
+    cannot_see = {}
     here_count = nil
 
     topic_id = params[:topic_id]
     if topic_id.present? && topic = Topic.find_by(id: topic_id)
       usernames.each do |username|
-        if !Guardian.new(User.find_by_username(username)).can_see?(topic)
-          cannot_see.push(username)
+        user = User.find_by_username(username)
+        next if user.blank?
+
+        cannot_see_reason = nil
+        if !Guardian.new(user).can_see?(topic)
+          cannot_see_reason = topic.private_message? ? :private : :category
+        elsif TopicUser.exists?(topic: topic, user: user, notification_level: TopicUser.notification_levels[:muted])
+          cannot_see_reason = :muted_topic
+        elsif topic.private_message? && !TopicAllowedUser.exists?(topic: topic, user: user) && !TopicAllowedGroup.exists?(topic: topic, group: user.groups)
```

Comment:
I think having test coverage for this line will be helpful since it involves permission checks. Also, I think you might be able to check whether a user has access to a topic by using the `Topic.private_messages_for_user` scope.